### PR TITLE
Update flatposts on more changes

### LINKED ISSUE #252: Editing icons/characters does not update flat views
Check on this. I don't recall if they do or not. I don't expect they would, skipping validations, but maybe they do.

### LINKED ISSUE #1460: Deleting replies does not update flat posts
Because after_save is stupid and doesn't cover deletes

### LINKED ISSUE #1645: More things should queue GenerateFlatPostJob
* Reply deletion
* Character replace affecting the post
* Icon replace affecting the post
* Icon swap (replacing the image) on posts where that icon is used

### PR DESCRIPTION
**Pull request by teceler**
Fixes #1645 
Fixes #1460 
Fixes #252

* Regenerates flatpost when reply is destroyed
* Regenerates flatpost when icon or icon keyword is changed
* Regenerates flatpost when character name or screenname is changed
* Regenerates flatpost when username is changed or user is archived

### CODE REVIEW

File: app/models/icon.rb [hunk 1]
```diff
@@ -73,6 +74,12 @@ def clear_icon_ids
     UpdateModelJob.perform_later(Reply.to_s, { icon_id: id }, { icon_id: nil }, audited_user_id)
   end
 
+  def update_flat_posts
+    return unless saved_change_to_url? || saved_change_to_keyword
```
Comment (teceler) [inline app/models/icon.rb:78]: Pretty sure that should have a question mark, which raises the question of why tests passed...

Comment (Marri) [reply app/models/icon.rb:78]: I think Rails magic things often have functions with and without the keyword (possibly if they're using a variable with a getter that returns the same value?) but support changing it to use the ?

File: app/models/user.rb [hunk 2]
```diff
@@ -159,4 +160,10 @@ def blocked_or_hidden_posts(keyword, post_user_ids, full_user_ids)
       post_ids
     end
   end
+
+  def update_flat_posts
+    return unless saved_change_to_username || saved_change_to_deleted
```
Comment (Marri) [inline app/models/user.rb:165]: ?s for consistency perhaps

File: app/models/user.rb [hunk 3]
```diff
@@ -159,4 +160,10 @@ def blocked_or_hidden_posts(keyword, post_user_ids, full_user_ids)
...
+    post_ids = (Post.where(user_id: id).pluck(:id) + Reply.where(user_id: id).select(:post_id).distinct.pluck(:post_id)).uniq
```
Comment (Marri) [inline app/models/user.rb:166]: Possibly faster to use PostAuthor here

Comment (teceler) [reply app/models/user.rb:166]: -- oh, yeah, PostAuthor probably would cover that

**Review by Marri — APPROVED** [review]

**Pull request merged**